segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/AppendProcessor.java

```diff
@@ -229,19 +226,23 @@ private void handleAppendResult(final Append append, Long newWriteOffset, Throwa
             } else {
                 if (conditionalFailed) {
                     log.debug("Conditional append failed due to incorrect offset: {}, {}", append, exception.getMessage());
-                    if (state != null) {
+                    synchronized (state.getAckLock()) {
                         // Revert the state to the last known good one. This is needed because we do not close the connection
                         // for offset-conditional append failures, hence we must revert the effects of the failed append.
-                        state.conditionalAppendFailed();
+                        state.conditionalAppendFailed(append.getEventNumber());
+                        connection.send(new ConditionalCheckFailed(append.getWriterId(), append.getEventNumber(), append.getRequestId()));
                     }
-                    connection.send(new ConditionalCheckFailed(append.getWriterId(), append.getEventNumber(), append.getRequestId()));
                 } else {
-                    // Clear the state in case of error.
-                    this.writerStates.remove(Pair.of(append.getSegment(), append.getWriterId()));
-                    handleException(append.getWriterId(), append.getRequestId(), append.getSegment(), append.getEventNumber(),
-                            "appending data", exception);
+                    // Record the exception handling into the Writer State. It will be executed  once all the current
```

Comment:
I have a concern with this model: It allows for the possibility that an append could fail but that one pipelined after it could succeed. Obviously this should not be possible and there is a check to try to prevent this. But if that logic for whatever reason had an edge case were it allowed that to occur, then this code would cause the ack of the later event to be sent to the client first effectively acking the failed append also. Can we incorporate some sort of assertion to prevent this possibility form occurring and hopefully aid in debugging it if it were to trigger?